### ISSUE TITLE
[captcha] Create captcha cog

### ISSUE THREAD

**Issue by calebj** [calebj/calebj-cogs#50]
As per Cog-Creators/Cog-Board#100

Planned features:
- [x] per-server config:
  - [x] designated captcha channel
  - [x] challenge generator, ~~length and charset~~
  - [x] approval or pending role modes
    - [x] automatically challenge when role is manually added?
  - [x] option to challenge on join, or just manually
  - [x] timeout (kick upon expiration)
- [x] plaintext, captcha and wheezy.captcha generators
- [x] user request of new challenge (with cooldown)
- [x] moderator manual and batch approval
- [x] no challenge for bots

Stretch goal(s):
- [x] ~~automatically unapprove command or link spammers~~ possible via exposed API

**Comment by skeith** [comment]
Since the original requester is MIA, I am willing to add $10 bounty if you can make it force captcha (make them added to non-approved / not-trusted role which can not see all other channel but one) on users that spam commands and/or spam links

If this idea is inefficient, please let me know

**Comment by calebj** [comment]
That's a good idea, but perhaps detecting spammers is a bit beyond the scope of this cog specifically. If there's already a cog out there that does what you want, but reacts by other means like a kick or ban (e.g. mentionspam in core mod), I can probably modify it to connect with this one.

**Comment by skeith** [comment]
Unfortunately, I don't know such cog yet

update:  
What if I change the bounty for making Guild Temperature cog (idea initially said by Kowlin back then).
Raise the temperature of a guild in case of suspicious activity and when the temp reaching a certain level, locks out the server, disabling invites and reporting suspicious users based on their activity


I know, this is a general and vague idea

**Comment by calebj** [comment]
Yeah, that's the scope of a separate cog. Though I'll be sure to build in the ability to trigger an (un)approval from outside the cog.

**Comment by calebj** [comment]
This is in the devlop branch, but I'll wait until I deem it fully stable to merge into master and close this.

@skeith there is an API for triggering a challenge. It's as simple as:
```py
await self.bot.get_cog('Captcha').challenge(bad_member)
```
from another cog, assuming the cog is loaded. If captcha is enabled in the server, it'll clear their role and act as if they're a new member.

**Comment by calebj** [comment]
Just a status update for anyone watching: this is still blocked by Cog-Creators/Red-DiscordBot#1927.
